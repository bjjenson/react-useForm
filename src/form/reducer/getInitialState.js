import { fromJS, Map } from 'immutable'
import { fieldsKey } from './fieldReducer'
import { generateDefaultListState } from './generateDefaultListState'
import { generateDefaultFieldState } from './generateDefaultFieldState'

/**
 * @param fields {Array<import("..").IFormFieldArgs>}
 */
export const getInitialState = (fields, initialValues = Map(), options = {}) => {
  const fieldMap = fields.reduce((acc, field) => {
    const result = acc.set(field.name, getFieldState(field, initialValues, options))

    return result
  }, Map())

  return fromJS({ [fieldsKey]: fieldMap })
}

const getFieldState = (field, initialValues, options) => {
  if (field.type === 'list') {
    return generateDefaultListState(field, initialValues, options)
  }
  return generateDefaultFieldState(field, initialValues, options)
}

