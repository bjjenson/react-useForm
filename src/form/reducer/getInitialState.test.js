import { fromJS } from 'immutable'
import { getInitialState } from './getInitialState'

jest.mock('react')
jest.useFakeTimers()

let fields, initialValues
beforeEach(() => {
  fields = [
    { name: 'first', label: 'First' },
    { name: 'second', label: 'Second', type: 'boolean' },
  ]
  initialValues = fromJS({
    first: 'firstValue',
    second: 'secondValue',
  })

})

test('prepares initialState by field', () => {
  expect(getInitialState(fields, initialValues)).toMatchSnapshot()
})

test('throws error if name not provided', () => {
  fields.push({ label: 'Third' })
  expect(() => getInitialState(fields, initialValues)).toThrowErrorMatchingSnapshot()
})

test('adds helperText from field config', () => {
  fields[0].helperText = 'i am helper'
  expect(getInitialState(fields, initialValues)).toMatchSnapshot()
})

test('sets optional on field', () => {
  fields[0].optional = true
  expect(getInitialState(fields, initialValues)).toMatchSnapshot()
})

test('sets optional on field using custom formatter', () => {
  const optionalLabelFormatter = jest.fn()
  optionalLabelFormatter.mockReturnValue('label formatted')
  fields[0].optional = true
  expect(getInitialState(fields, initialValues, { optionalLabelFormatter })).toMatchSnapshot()

  expect(optionalLabelFormatter.mock.calls[0]).toMatchSnapshot()
})

test('reads initialValues that are nested', () => {
  fields = [
    { name: 'nested.first', label: 'First' },
  ]
  initialValues = fromJS({
    nested: {
      first: 'nestedValue',
    },
  })
  expect(getInitialState(fields, initialValues)).toMatchSnapshot()
})

describe('initialValues', () => {
  test('field.value should used if set', () => {
    fields = [
      { name: 'first', type: 'text', value: 'i am field value' },
    ]
    expect(getInitialState(fields)).toMatchSnapshot()
  })

  test('text', () => {
    fields = [
      { name: 'first', type: 'text' },
    ]
    expect(getInitialState(fields)).toMatchSnapshot()
  })

  test('select', () => {
    fields = [
      { name: 'first', type: 'select' },
    ]
    expect(getInitialState(fields)).toMatchSnapshot()
  })

  test('number', () => {
    fields = [
      { name: 'first', type: 'number' },
    ]
    expect(getInitialState(fields)).toMatchSnapshot()
  })

  test('boolean', () => {
    fields = [
      { name: 'first', type: 'boolean' },
    ]
    expect(getInitialState(fields)).toMatchSnapshot()
  })

  test('list', () => {
    fields = [
      { name: 'first', type: 'list' },
    ]
    expect(getInitialState(fields)).toMatchSnapshot()
  })
})
