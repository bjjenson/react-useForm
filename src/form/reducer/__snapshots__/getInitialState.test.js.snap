// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds helperText from field config 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "text",
        "value": "firstValue",
        "optional": false,
        "label": "First",
        "field": Object {
          "helperText": "i am helper",
          "label": "First",
          "name": "first",
        },
      },
      "current": Immutable.Map {
        "helperText": "i am helper",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "firstValue",
      },
    },
    "second": Immutable.Map {
      "initial": Immutable.Map {
        "type": "boolean",
        "value": "secondValue",
        "optional": false,
        "label": "Second",
        "field": Object {
          "label": "Second",
          "name": "second",
          "type": "boolean",
        },
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "secondValue",
      },
    },
  },
}
`;

exports[`initialValues boolean 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "boolean",
        "value": false,
        "optional": false,
        "label": undefined,
        "field": Object {
          "name": "first",
          "type": "boolean",
        },
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": false,
      },
    },
  },
}
`;

exports[`initialValues field.value should used if set 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "text",
        "value": "i am field value",
        "optional": false,
        "label": undefined,
        "field": Object {
          "name": "first",
          "type": "text",
          "value": "i am field value",
        },
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "i am field value",
      },
    },
  },
}
`;

exports[`initialValues list 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "list",
        "optional": false,
        "label": undefined,
        "field": Object {
          "name": "first",
          "type": "list",
        },
        "options": Object {},
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
      },
      "items": Immutable.List [],
    },
  },
}
`;

exports[`initialValues number 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "number",
        "value": 0,
        "optional": false,
        "label": undefined,
        "field": Object {
          "name": "first",
          "type": "number",
        },
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": 0,
      },
    },
  },
}
`;

exports[`initialValues select 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "select",
        "value": "",
        "optional": false,
        "label": undefined,
        "field": Object {
          "name": "first",
          "type": "select",
        },
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "",
      },
    },
  },
}
`;

exports[`initialValues text 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "text",
        "value": "",
        "optional": false,
        "label": undefined,
        "field": Object {
          "name": "first",
          "type": "text",
        },
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "",
      },
    },
  },
}
`;

exports[`prepares initialState by field 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "text",
        "value": "firstValue",
        "optional": false,
        "label": "First",
        "field": Object {
          "label": "First",
          "name": "first",
        },
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "firstValue",
      },
    },
    "second": Immutable.Map {
      "initial": Immutable.Map {
        "type": "boolean",
        "value": "secondValue",
        "optional": false,
        "label": "Second",
        "field": Object {
          "label": "Second",
          "name": "second",
          "type": "boolean",
        },
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "secondValue",
      },
    },
  },
}
`;

exports[`reads initialValues that are nested 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "nested.first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "text",
        "value": "nestedValue",
        "optional": false,
        "label": "First",
        "field": Object {
          "label": "First",
          "name": "nested.first",
        },
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "nestedValue",
      },
    },
  },
}
`;

exports[`sets optional on field 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "text",
        "value": "firstValue",
        "optional": true,
        "label": "First (optional)",
        "field": Object {
          "label": "First",
          "name": "first",
          "optional": true,
        },
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "firstValue",
      },
    },
    "second": Immutable.Map {
      "initial": Immutable.Map {
        "type": "boolean",
        "value": "secondValue",
        "optional": false,
        "label": "Second",
        "field": Object {
          "label": "Second",
          "name": "second",
          "type": "boolean",
        },
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "secondValue",
      },
    },
  },
}
`;

exports[`sets optional on field using custom formatter 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "text",
        "value": "firstValue",
        "optional": true,
        "label": "label formatted",
        "field": Object {
          "label": "First",
          "name": "first",
          "optional": true,
        },
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "firstValue",
      },
    },
    "second": Immutable.Map {
      "initial": Immutable.Map {
        "type": "boolean",
        "value": "secondValue",
        "optional": false,
        "label": "Second",
        "field": Object {
          "label": "Second",
          "name": "second",
          "type": "boolean",
        },
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "secondValue",
      },
    },
  },
}
`;

exports[`sets optional on field using custom formatter 2`] = `
Array [
  "First",
  "first",
]
`;

exports[`throws error if name not provided 1`] = `"name is required on field"`;
